src/common/ipc.ts

```diff
@@ -2,94 +2,133 @@
 // https://www.electronjs.org/docs/api/ipc-main
 // https://www.electronjs.org/docs/api/ipc-renderer
 
-import { ipcMain, ipcRenderer, WebContents, webContents } from "electron"
+import { ipcMain, ipcRenderer, IpcRendererEvent, WebContents, webContents } from "electron"
 import logger from "../main/logger";
+import { getRandId } from "./utils";
 
 export type IpcChannel = string;
 
-export interface IpcHandleOpts {
-  timeout?: number;
+export enum IpcMode {
+  SYNC = "sync",
+  ASYNC = "async",
 }
 
-export interface IpcMessageHandler<T extends any[] = any> {
-  (...args: T): any;
+export interface IpcChannelRequest<A extends any[] = any> {
+  msgId: string;
+  args: A;
 }
 
-export interface IpcMessageOpts<A extends any[] = any> {
+export interface IpcChannelResponse<T extends any[] = any, E = any> {
+  msgId: string;
+  data?: T;
+  error?: E;
+}
+
+export interface IpcChannelInit {
+  channel: IpcChannel; // main <-> renderer communication channel name
+  mode?: IpcMode; // default: "async", use "sync" as last resort: https://www.electronjs.org/docs/api/ipc-renderer#ipcrenderersendsyncchannel-args
+  handle?: (...args: any[]) => any; // main-process message handler
+  autoBind?: boolean; // auto-bind message handler in main-process, default: false
+  timeout?: number; // timeout for waiting response from the sender
+  once?: boolean; // todo: add support
+}
+
+export function createIpcChannel({ autoBind = true, mode = IpcMode.ASYNC, timeout = 0, handle, channel }: IpcChannelInit) {
+  channel = `${mode}:${channel}`
+
+  const ipcChannel = {
+    channel: channel,
+    handleInMain: () => {
+      logger.info(`[IPC]: setup channel "${channel}"`);
+
+      ipcMain.on(channel, async (event, req: IpcChannelRequest) => {
+        let resolved = false;
+        let timerId: any;
+
+        function resolve(res: Partial<IpcChannelResponse>) {
+          if (resolved) return;
+          res.msgId = req.msgId; // return back to sender to be able to handle response
+          resolved = true
+          logger.debug(`[IPC]: sending response to "${channel}"`, res);
+          if (mode === IpcMode.ASYNC) {
+            event.reply(channel, res);
+          }
+          if (mode === IpcMode.SYNC) {
+            event.returnValue = res;
+          }
```

Comment:
shouldn't this either be a `switch` or `if ... else if`?

Reply:
doesn't really matter, only if it's a question of code-readability? :)

Reply:
IMHO `if ... else if` would be better here.